### PR TITLE
[AMDAIEDistributeCoresAndObjectFifos] Factorize out memory privatization sub-pass

### PR DESCRIPTION
**Pull request by newling**
By running this sub-pass earlier, it can be made simpler because it doesn't need to handle anything DMA related. Also this sub-pass is quite modular and doesn't need to be part of the larger distribution pass. My initial motivation for this factorization was that it makes running aievec passes before the cores are materialized easier (so can run aievec passes on just a single function, instead of 16 cores).

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    ArrayRef<int64_t> newShape = memRefType.getShape();
+    Type elementType = memRefType.getElementType();
+
+    rewriter.setInsertionPoint(oldAlloc);
+    MemRefType newAllocType = MemRefType::get(
+        newShape, elementType, MemRefLayoutAttrInterface{}, memorySpace);
+    auto newAlloc = rewriter.create<memref::AllocOp>(rewriter.getUnknownLoc(),
+                                                     newAllocType);
+
+    auto users_ = oldAlloc->getUsers();
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp:126]: Trailing underscores are widely used for class member variables, so I think we should avoid using it.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+          llvm::TypeSwitch<Operation *, LogicalResult>(user)
+              .Case<memref::SubViewOp>([&](memref::SubViewOp subviewOp) {
+                rewriter.replaceAllUsesWith(subviewOp, newAlloc);
+                return success();
+              })
+              .Case<vector::TransferReadOp>([&](vector::TransferReadOp readOp) {
+                rewriter.setInsertionPoint(readOp);
+                Value c0 =
+                    rewriter.create<arith::ConstantIndexOp>(readOp.getLoc(), 0);
+                auto indices = substitute(readOp.getIndices(), indVars, c0);
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp:141]:
```suggestion
                SmallVector<Value> indices = substitute(readOp.getIndices(), indVars, c0);
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                    readOp.getPermutationMapAttr(), readOp.getPadding(),
+                    readOp.getMask(), readOp.getInBoundsAttr());
+                return success();
+              })
+              .Case<vector::TransferWriteOp>(
+                  [&](vector::TransferWriteOp writeOp) {
+                    rewriter.setInsertionPoint(writeOp);
+                    Value c0 = rewriter.create<arith::ConstantIndexOp>(
+                        writeOp.getLoc(), 0);
+                    auto indices =
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp:153]:
```suggestion
                    SmallVector<Value> indices =
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      Operation::operand_range offsets = subview.getOffsets();
+      for (Value offset : offsets) {
+        if (matchPattern(offset, m_Constant()))
+          continue;
+        else if (llvm::is_contained(indVars, offset))
+          continue;
+        // We assume that if a subview has an offset which is not
+        // a constant and not a thread id, it's not 'distributing'.
+        else
+          return {};
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp:61]: Use curly braces: https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/distribute_l1_allocations.mlir [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    linalg.fill ins(%c0_i32 : i32) outs(%subview : memref<1x1x32x32xi32, strided<[2048, 1024, 32, 1], offset: ?>, 2>)
+  } {mapping = [#gpu.thread<y>, #gpu.thread<x>]}
+  memref.dealloc %alloc : memref<2x2x32x32xi32, 2>
+  return
+}
+
+// -----
+
+
+// TODO(newling) add more tests.
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/distribute_l1_allocations.mlir:36]: It's your intention to do this in this PR, right?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  moduleOp.walk([&](scf::ForallOp forallOp) {
+    auto maybeMapping = forallOp.getMapping();
+    if (!maybeMapping) return WalkResult::advance();
+    SmallVector<Attribute> mapping = llvm::to_vector(maybeMapping->getValue());
+    if (mapping.empty()) return WalkResult::advance();
+    if (!isa<mlir::gpu::GPUThreadMappingAttr>(*mapping.begin()))
+      return WalkResult::advance();
+    for (auto indVar : forallOp.getInductionVars()) {
+      threadIndVars.insert(indVar);
+    }
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp:39]:
```suggestion
    for (auto indVar : forallOp.getInductionVars())
      threadIndVars.insert(indVar);
```

**Review by jtuyls — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp (hunk 4 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeL1Allocations.cpp]: I simplified, removed continue statements

**Review by jtuyls — APPROVED** [review]
LGTM

**Pull request merged**